### ISSUE TITLE
Can peft support singular value decomposition?

### ISSUE THREAD

**Issue by hobbitlzy** [huggingface/peft#1270]
Hi, I am new to `peft`. I am wondering whether the `peft` supports use singular value decomposition to replace the orignal linear module. I find this is similar to the funciton of lora, but different at the `loar_A` and `lora_B` are decomposition of the original weights. 

If this does not support, is there any advice to implement such functionality?

**Comment by BenjaminBossan** [comment]
As you mention, this is related to the idea of LoRA, but I'm not aware of any fine-tuning technique that consists of decomposing the original weight. Intuitively, this would lead to a big loss in model performance. If there are any papers that show that this works for fine-tuning, please post them.

**Comment by hobbitlzy** [comment]
It's just one attempt, and I've observed that directly decomposing the trained weights results in a significant loss. I am interested in exploring whether such a loss can be rectified through fine-tuning.

**Comment by pacman100** [comment]
Hello @hobbitlzy,

Please see this paper https://arxiv.org/pdf/2209.05310v1.pdf wherein one can convert the existing linear layers into low rank matrices by doing SVD decomposition. This paper does it (section 3.1 on Bottlenecks), result of it: reduced time per training step by 7% without impacting accuracy

**Comment by EeyoreLee** [comment]
The lightweight maybe the literal light weight Xd, like Albert embedding. No reduction in computational effort but save a little memory. And if it is a feasible solution, I think Albert won't just factorizate the embedding.

**Comment by BenjaminBossan** [comment]
> Instead, I am curious about if SVD can be applied to a pre-trained model in a way that enables lightweight fine-tuning to compensate for any performance degradation.

To be clear, do you suggest that a pretrained model is taken, its big weight matrices decomposed using SVD, then LoRA is added on top to compensate for the error introduced by the decomposition? Intuitively, I would imagine that this is not too efficient, since LoRA is also a kind of decomposition, so I believe adding it would just have a very similar effect as increasing the rank of the SVD.

**Comment by pacman100** [comment]
> Instead, I am curious about if SVD can be applied to a pre-trained model in a way that enables lightweight fine-tuning to compensate for any performance degradation.

This is unclear to me. Can you clarify this further in terms of pseudocode that you have in mind?

**Comment by BenjaminBossan** [comment]
My understanding is that this idea is similar to LoftQ, where the LoRA weights are used to compensate for errors introduced by quantization, here it could be used to compensate for the error from factorizing the matrix. But I think it makes less sense here, since the two operations (SVD and LoRA) are so similar that the latter cannot really compensate for the former.

**Comment by hobbitlzy** [comment]
Apologies for the lack of clarity earlier. Actually, my objective is to apply SVD to compress a pre-trained model, followed by a fine-tuning process to restore its accuracy. This fine-tuning could be conventional or could utilize LoRA. The term 'lightweight' was used in comparison to certain methods that use SVD to train a model from scratch.

Furthermore, I agree that incorporating LoRA into this fine-tuning phase makes less sense, as it is equivalent to choosing a higher SVD rank.

**Comment by EeyoreLee** [comment]
@hobbitlzy hi, I'm confused what's the purpose to factorize the  parameters matrix. Just for less storage on disk ? it won't be save more gpu memory and won't be faster than lora and other peft.

**Comment by hobbitlzy** [comment]
@EeyoreLee  Hi, the target scenario for me is privacy-preserving machine learning and it indeed has improvements on both memory and computation.

**Comment by EeyoreLee** [comment]
@hobbitlzy - it indeed save mermory only on inference and it won't save computation on both training and inference.  For model training step, the memory is cost on optim, parameters, gradient and activation value. Especially those optimizers that use both first-order and second-order momentum like Adam.

Let’s first discuss memory overhead. Although the rank matrix factorizing reduce memory consumption on parameters, it will bring more gradients and corresponding optimizer consumption than LoRA since LoRA only need gradient and its corresponding optimizer parameters for A matrix and B matrix

Second for computation, if the origin computation is matrix A(m x n) dot matrix B(n x q) and we SVD B to XYZ which X(n x p), Y(p x r), Z(r x q), the compustation is changed from O(mnq) to O(mnp)+O(mpr)+O(mrq). Unless the p and r is much smaller than n and q, it even bring more computation. And to maintain as much accuracy as possible, the p and r is impossible to small enough.

**Comment by hobbitlzy** [comment]
Thanks for the patient explanation. I understand your point except a slight disagreement for the computation overhead at the inference stage. 

> Second for computation, if the origin computation is matrix A(m x n) dot matrix B(n x q) and we SVD B to XYZ which X(n x p), Y(p x r), Z(r x q), the compustation is changed from O(mnq) to O(mnp)+O(mpr)+O(mrq). Unless the p and r is much smaller than n and q, it even bring more computation. And to maintain as much accuracy as possible, the p and r is impossible to small enough.

In my understanding, the B is decomposed as XYZ, where Y is a diagnal matrix and we only keep the top `r` values. After tha Y is absorbed into either X or Y. Then B(n x q) is trasnformed as X(n x r) and Y(r x q). THe computation is changed from O(mnq) to O(mnr)+O(mrq). If `r` is small, it indeeds reduce the comptuation complexity.

**Comment by EeyoreLee** [comment]
@hobbitlzy - It's my turn to say apologize for thinking too decisively lol. you're right that seems if r smaller than half of n or q, the comptuation complexity is indeed smaller in big O notation, only may need little extra time cost on transfering value from memory to register. Looks like a feasible solution. The most important thing should verify is whether it can hold the accuracy and get a small enough `r` at a time.

**Comment by hobbitlzy** [comment]
@EeyoreLee Glad to reach the agreement. The concern in my scenario is the computation overhead at the runtime (although can be quite difference in the private computation). But I do not find any related codes yet and this may be implemented myself.

**Comment by BenjaminBossan** [comment]
This method could be interesting to investigate in this context: https://github.com/pratyushasharma/laser
